### ISSUE TITLE
How to process extracted values?

### ISSUE THREAD

**Issue by HelgeSverre** [crwlrsoft/crawler#75]
Great library, love the simplicity, however I have a question related to processing the extracted data that is not obvious reading the docs.

Example:

```PHP
$crawler = new Crawler();
$crawler
    ->setStore(new JsonFileStore(base_path(), "output.json"))
    ->input('http://books.toscrape.com/')
    ->addStep(Http::get())
    ->addStep(Html::getLinks('.product_pod .image_container a'))
    ->addStep(Http::get())
    ->addStep(
        Html::first("article")->extract([
            'title' => 'h1',
            'price' => Dom::cssSelector('.price_color')->first(),
            'stock' => Dom::cssSelector('.availability')->first(),
        ])->addToResult()
    );

foreach ($crawler->run() as $result) {
    dump($result->toArray());
    
    // [
    //      "title" => "Rip it Up and Start Again"
    //      "price" => "£35.02"
    //      "stock" => "In stock (19 available)" <--- How would i go about extracting only the number as part of the extraction?
    // ]
}
```

Where and how is the appropriate place to "parse" the value of the "stock" field to extract the number itself, the value cannot be grabbed by selector only, as the number value itself is not surrounded by a tag.

It might "belong" in the custom Store, where one would "parse" the extracted data further, but for things like this it would be convenient to be able to do something like this (Str in the example is the Laravel Stringable class):

```php
Html::first("article")->extract([
    'title' => 'h1',
    'price' => Dom::cssSelector('.price_color')->first(),
    'stock' => Dom::cssSelector('.availability')->first()->value(
        callback: fn($value) => Str::of($value)->after("(")->before(" ")->toString()
    ),
])->addToResult()
```

Is there currently a way to do something similar?

**Comment by otsch** [comment]
Hey @HelgeSverre, I'm very glad you like the library :)

About your question: I've actually just recently planned to add new functionality for this 😅. I'm working on v1.0 this week and will add it. I'm thinking of something like a `refineOutput()` method that you can use with any step. For your example, like:

```php
Html::first("article")
    ->extract([
        'title' => 'h1',
        'price' => Dom::cssSelector('.price_color')->first(),
        'stock' => Dom::cssSelector('.availability')->first(),
    ])
    ->refineOutput(function ($output, $input) {
        $output['stock'] = Str::of($output['stock'])->after("(")->before(" ")->toString();

        return $output;
    })
    ->addToResult()
```

I'd prefer to solve it this way, because as mentioned you'll be able to use this with any step.

**Comment by HelgeSverre** [comment]
Looks like a clean solution, keep up the great work :)

**Comment by otsch** [comment]
Hey @HelgeSverre 👋 if you like to, you can already try it from my v1.0 dev branch: https://packagist.org/packages/crwlr/crawler#dev-v1-0 
```bash
composer require crwlr/crawler:dev-v1-0
```

You can refine array output using one `Closure` like already shown above: 
```php
Html::first("article")
    ->extract([
        'title' => 'h1',
        'price' => Dom::cssSelector('.price_color')->first(),
        'stock' => Dom::cssSelector('.availability')->first(),
    ])
    ->refineOutput(function ($output, $input) {
        $output['stock'] = Str::of($output['stock'])->after("(")->before(" ")->toString();

        return $output;
    })
    ->addToResult()
```

You can also refine single array output elements by key, by providing the key as first argument to the `refineOutput` method:
```php
Html::first("article")
    ->extract([
        'title' => 'h1',
        'price' => Dom::cssSelector('.price_color')->first(),
        'stock' => Dom::cssSelector('.availability')->first(),
    ])
    ->refineOutput('stock', fn ($output, $input) => Str::of($output)->after("(")->before(" ")->toString())
    ->addToResult()
```

And I've even added a few simple "Refiners" that you can pass to the method instead of a Closure, like:
```php
Html::first("article")
    ->extract([
        'title' => 'h1',
        'price' => Dom::cssSelector('.price_color')->first(),
        'stock' => Dom::cssSelector('.availability')->first(),
    ])
    ->refineOutput('stock', Str::betweenFirst('(', ' '))
    ->addToResult()
```

The `Str` class providing string Refiners via static methods is: `Crwlr\Crawler\Steps\Refiners\Str`.
I'm just thinking, maybe I'll change it before v1.0 release, so naming doesn't conflict with the laravel string helper facade 🤔 maybe to something like `StringRefiner`.
Hope you like it :)

**Comment by HelgeSverre** [comment]
@otsch Will check it out later tonight!

**Comment by HelgeSverre** [comment]
Seems to work great!

```php

$crawler = new Crawler();
$crawler
    ->setStore(new JsonFileStore(base_path(), "output.json"))
    ->input('http://books.toscrape.com/')
    ->addStep(Http::get())
    ->addStep(Html::getLinks('.product_pod .image_container a'))
    ->addStep(Http::get())
    ->addStep(
        Html::first("article")
            ->extract([
                'title' => 'h1',
                'price' => Dom::cssSelector('.price_color')->first()->text(),
                'stock' => Dom::cssSelector('.availability')->first(),
            ])
            ->refineOutput(function ($output, $input) {
                $output['stock'] = Str::of($output['stock'])->after("(")->before(" ")->toInteger();
                $output['price'] = Str::of($output['price'])->remove("£")->toFloat();

                return $output;
            })
            ->addToResult()
    );

foreach ($crawler->run() as $result) {
    dump($result->toArray());
}
```

**Comment by otsch** [comment]
Thanks again @HelgeSverre 👍🏻 
I close this issue now, as the feature is now included in v1.0.0 that I released today 🙂 
https://www.crwlr.software/packages/crawler/v1.0/steps-and-data-flow/refiners
